### PR TITLE
Update ssh keys and user name using CLI

### PR DESCRIPTION
**Pull request by mitalipaygude**
*Issue #, if available:*
#1208 

*Description of changes:*
These code changes allows an user to update the ssh keys and the user name for vsphere machine configs.
If the user updates the `etcd` or the `control plane` machine configs, we update the `template-cp` and this change triggers CAPI to roll out new nodes.
In case of worker node machine config updates, we create a new `kubeadm config template` and update the worker nodes with the new user/ssh keys.

*Testing (if applicable):*
- Upgrade management cluster's `etcd`, `control plane` and `worker node` machine configs with new users and ssh keys using `upgrade cluster` (CLI only). Logged into the vms with the updated keys.
- Upgrade workload cluster's `etcd`, `control plane` and `worker node` machine configs with new users and ssh keys using `upgrade cluster` (CLI only). Logged into the vms with the updated keys

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: controllers/controllers/resource/reconciler_test.go [hunk 1]
```diff
@@ -152,13 +152,23 @@ func TestClusterReconcilerReconcile(t *testing.T) {
 				if err := yaml.Unmarshal([]byte(machineDeploymentFile), machineDeployment); err != nil {
 					t.Errorf("unmarshalling machinedeployment failed: %v", err)
 				}
+				workerNodeMachineConfig := &anywherev1.VSphereMachineConfig{
```
Comment (jiayiwang7) [inline controllers/controllers/resource/reconciler_test.go:155]: this pr also covers controller updates? if not fully, let's maybe move the controller changes into separate pr

Comment (mitalipaygude) [reply controllers/controllers/resource/reconciler_test.go:155]: There is common code that calls the adding new kubeadm config template, which is why these changes were necessary. However the immutable changes for ssh keys in the webhook haven't been updated, so the updates to the ssh keys wont go through.

File: pkg/providers/vsphere/vsphere.go [hunk 2]
```diff
@@ -1182,23 +1182,23 @@ func (p *vsphereProvider) ValidateNewSpec(ctx context.Context, cluster *types.Cl
 	return nil
 }
 
-func (p *vsphereProvider) needsNewMachineTemplate(ctx context.Context, workloadCluster *types.Cluster, currentSpec, newClusterSpec *cluster.Spec, workerNodeGroupConfiguration v1alpha1.WorkerNodeGroupConfiguration, vdc *v1alpha1.VSphereDatacenterConfig, prevWorkerNodeGroupConfigs map[string]v1alpha1.WorkerNodeGroupConfiguration) (bool, error) {
+func (p *vsphereProvider) needsNewMachineTemplate(ctx context.Context, workloadCluster *types.Cluster, currentSpec, newClusterSpec *cluster.Spec, workerNodeGroupConfiguration v1alpha1.WorkerNodeGroupConfiguration, vdc *v1alpha1.VSphereDatacenterConfig, prevWorkerNodeGroupConfigs map[string]v1alpha1.WorkerNodeGroupConfiguration) (bool, error, *v1alpha1.VSphereMachineConfig, *v1alpha1.VSphereMachineConfig) {
```
Comment (jiayiwang7) [inline pkg/providers/vsphere/vsphere.go:1185]: this method should not return objects

Comment (mitalipaygude) [reply pkg/providers/vsphere/vsphere.go]: Added new method

File: pkg/api/v1alpha1/vspheremachineconfig_types.go [hunk 3]
```diff
@@ -20,6 +20,25 @@ type VSphereMachineConfigSpec struct {
 	Users             []UserConfiguration `json:"users,omitempty"`
 }
 
+func UsersSliceEqual(a, b []UserConfiguration) bool {
```
Comment (jiayiwang7) [inline pkg/api/v1alpha1/vspheremachineconfig_types.go:23]: can we add unit test?

**Review by jiayiwang7 — APPROVED** [review]

**Comment by mitalipaygude** [comment]
#1440

File: pkg/api/v1alpha1/vspheremachineconfig_types.go (hunk 3 above)
Comment (mitalipaygude) [reply pkg/api/v1alpha1/vspheremachineconfig_types.go:23]: #1440
